kv/dist_sender.go

```diff
@@ -397,6 +397,50 @@ func (ds *DistSender) Send(call *client.Call) {
 			} else if isMulti {
 				// If this request spans ranges, collect the replies.
 				responses = append(responses, reply)
+
+				// If this request is bounded with a positive MaxResults,
+				// check whether the existing rows are enough.
+				var maxResults int64
+				if args, ok := args.(proto.Bounded); ok && args.GetBound() > 0 {
+					maxResults = args.GetBound()
+					if _, ok := responses[0].(proto.Truncatable); ok {
+						var accumulatedResults int64
+						// Calculate the number of the existing rows.
+						for _, v := range responses {
+							accumulatedResults += v.(proto.Truncatable).RowCount()
+						}
+						if accumulatedResults >= maxResults {
+							// Set flag to break loop.
+							descNext = nil
+						}
+					}
+				}
+
+				// descNext can be nil in two cases:
+				// 1. Got enough rows in the middle of the request.
+				// 2. It is the last range of the request.
+				// Then combine and truncate the responses.
+				if descNext == nil {
+					// Combine multiple responses into the first one.
+					if len(responses) > 1 {
+						// We've already ascertained earlier that we're dealing with a
+						// Combinable response type.
+						firstReply := responses[0].(proto.Combinable)
+						for _, r := range responses[1:] {
+							firstReply.Combine(r)
+						}
+					}
+
+					// Truncate the response if MaxResults is specified.
+					if maxResults > 0 {
```

Comment:
Another option which might be slightly more efficient is to keep track of the running total in the outer loop (i.e. move accumulatedResults var to a higher-level scope) and provide a SetBound() method on the proto.Bounded interface for each call through to a range.